### PR TITLE
[RFC] libsks for Pkcs#11 services through OP-TEE SKS TA

### PR DESCRIPTION
**Pull request by etienne-lms**
This change introduces an implementation of the Pkcs#11 Cryptoki library that interface a specific trusted application in the secure side provided by the optee_os source tree.

This P-R implements the basics for the client library. The effective implementation of the services will come later. However this P-R still introduces a first TA command for this initial P-R to at least invoke the SKS TA to check it is available when library is initialized. 

This P-R does not introduce the full TA API itself that is being reviewed through (**edited: https://github.com/OP-TEE/optee_os/pull/2732**). The client library will integrate the released TA API once its is approved and merged in optee_os mainline.

This P-R replaces #121.

**edited, url of TA API P-R and below**
https://github.com/OP-TEE/optee_client/pull/139 proposes the first commits of this RFC.

### CODE REVIEW

File: libsks/src/invoke_ta.c [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+{
+	return (TEEC_Context *)sks_ctx->context;
+}
+
+static TEEC_Session *teec_sess(struct sks_invoke *sks_ctx)
+{
+	return (TEEC_Session *)sks_ctx->session;
+}
+
+TEEC_SharedMemory *sks_alloc_shm(struct sks_invoke *sks_ctx,
```
Comment (etienne-lms) [inline libsks/src/invoke_ta.c]: I will remove this invocation context argument. All invocation should use the TEE context/session created from `open_primary_context()`.

File: libsks/include/sks_ta.h [hunk 2]
```diff
@@ -0,0 +1,116 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2018, Linaro Limited
+ */
+
+#ifndef __SKS_TA_H__
```
Comment (wamserma) [inline libsks/include/sks_ta.h]: No code duplication here. This has to be pulled in from `optee_os`. Otherwise I'll bet the repos will get out-of-sync within one year, because people forget they have to make two PRs or one PR gets stalled.

Comment (jenswi-linaro) [reply libsks/include/sks_ta.h]: This is supposed to be a stable ABI so if it's out of sync or not shouldn't be a problem.

**Comment by etienne-lms** [comment]
Minor question but to be clarified asap if possible: since removing the SKS meaningless acronym, I'm looking for a nice name for the client library. Not too small, not too long, to address a cryptoki integration ina GTD TEE env (as OP-TEE is).
I am thinking of `libcryptokioptee`, `libckteec`, `libckoptee`.  Any other artistic idea?

File: libsks/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,72 @@
+project(sks C)
+
+set(PROJECT_VERSION "1.0.0")
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libsks/CMakeLists.txt]: Ok for `0.1.0`.

File: libsks/Makefile [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+all: libsks
+install: libsks
+
+LIB_NAME	:= libsks
+MAJOR_VERSION	:= 0
+MINOR_VERSION	:= 0
+
+LIB_MAJOR		:= $(LIB_NAME).so.$(MAJOR_VERSION)
+LIB_MAJ_MIN		:= $(LIB_NAME).so.$(MAJOR_VERSION).$(MINOR_VERSION)
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libsks/Makefile]: ok

File: libsks/include/pkcs11.h [hunk 5]
```diff
@@ -0,0 +1,1377 @@
...
+};
+
+/*
+ * Values for CK_ATTRIBUTE_TYPE
+ *
+ * This does not cover the full PKCS#11 IDs.
+ */
+#define CKF_ARRAY_ATTRIBUTE		(1U << 30)
+#define CKA_VENDOR_DEFINED		(1U << 31)
+#define CKA_CLASS			0x0000
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libsks/include/pkcs11.h]: no strong opinion.
Ok to move to an enumerate.

File: libsks/src/pkcs11_api.c [hunk 6]
```diff
@@ -0,0 +1,1123 @@
...
+	NOT_SUPPORTED_CK_FUNCTION(C_UnwrapKey),
+	NOT_SUPPORTED_CK_FUNCTION(C_DeriveKey),
+	NOT_SUPPORTED_CK_FUNCTION(C_SeedRandom),
+	NOT_SUPPORTED_CK_FUNCTION(C_GenerateRandom),
+	NOT_SUPPORTED_CK_FUNCTION(C_GetFunctionStatus),
+	NOT_SUPPORTED_CK_FUNCTION(C_CancelFunction),
+	NOT_SUPPORTED_CK_FUNCTION(C_WaitForSlotEvent),
+};
+
+static int lib_initiated;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libsks/src/pkcs11_api.c]: ok, done

File: libsks/src/pkcs11_api.c [hunk 7]
```diff
@@ -0,0 +1,1123 @@
...
+
+/*
+ * List of all PKCS#11 cryptoki API functions implemented
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libsks/src/pkcs11_api.c]: copy/paste issue. Thanks, fixed.

File: libsks/src/pkcs11_api.c [hunk 8]
```diff
@@ -0,0 +1,1123 @@
...
+	/* Argument currently unused */
+	(void)args;
+
+	if (lib_initiated)
+		return CKR_CRYPTOKI_ALREADY_INITIALIZED;
+
+	if (sks_invoke_init())
+		return CKR_FUNCTION_FAILED;
+
+	lib_initiated = 1;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libsks/src/pkcs11_api.c]: done

File: libsks/src/pkcs11_api.c [hunk 9]
```diff
@@ -0,0 +1,1123 @@
...
+CK_RV C_Finalize(CK_VOID_PTR res)
+{
+	(void)res;
+
+	if (!lib_initiated)
+		return CKR_CRYPTOKI_NOT_INITIALIZED;
+
+	sks_invoke_terminate();
+
+	lib_initiated = 0;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libsks/src/pkcs11_api.c]: done

File: libsks/src/pkcs11_api.c [hunk 10]
```diff
@@ -0,0 +1,1123 @@
...
+
+	/* Discard the const attribute when exporting the list address */
+	*ppFunctionList = (void *)&libsks_function_list;
+
+	return CKR_OK;
+}
+
+CK_RV C_GetSlotList(CK_BBOOL token_present,
+		    CK_SLOT_ID_PTR slots,
+		    CK_ULONG_PTR count)
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libsks/src/pkcs11_api.c]: Note really. Maybe because i am not found of camelcase style, but that shouldn't count :)

Ok to stick to the API for this file and the related header file.

**Comment by etienne-lms** [comment]
Refer to https://github.com/OP-TEE/optee_test/pull/281#issuecomment-576367540 for running the basic test.

**Comment by jenswi-linaro** [comment]
I'd prefer `libckteec`, but I don't have a strong opinion on this.

**Comment by etienne-lms** [comment]
let's go for `libckteec`. short and still refers to both `libteec` and `cryptoki`/`ck`.

File: libsks/src/pkcs11_api.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libsks/src/pkcs11_api.c]: addressed.

File: libsks/include/pkcs11.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libsks/include/pkcs11.h]: Not addressed yet!

File: libckteec/include/pkcs11.h [hunk 11]
```diff
@@ -0,0 +1,1378 @@
...
+
+/* Argument for C_GetFunctionList */
+typedef struct CK_FUNCTION_LIST		CK_FUNCTION_LIST;
+typedef struct CK_FUNCTION_LIST *	CK_FUNCTION_LIST_PTR;
+typedef struct CK_FUNCTION_LIST **	CK_FUNCTION_LIST_PTR_PTR;
+
+struct CK_FUNCTION_LIST {
+	CK_VERSION version;
+	CK_RV (*C_Initialize)(
+			CK_VOID_PTR pInitArgs);
```
Comment (jenswi-linaro) [inline libckteec/include/pkcs11.h]: The indentation looks a bit funny in this file. Did you copy it from somewhere? In that case please add a source in the commit message. If not, please fix all the indentation in this file.

File: Makefile [hunk 12]
```diff
@@ -35,18 +35,26 @@ build-tee-supplicant: build-libteec
...
 clean-libteec:
 	@$(MAKE) --directory=libteec --no-print-directory clean
 
 clean-tee-supplicant:
 	@$(MAKE) --directory=tee-supplicant --no-print-directory clean
 
+clean-libckteec:
+	@$(MAKE) --directory=libckteec --no-print-directory clean
+
+
```
Comment (jenswi-linaro) [inline Makefile]: Empty line not needed

File: libckteec/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+)
+
+target_include_directories(teec
+	PUBLIC include
+)
+
+target_link_libraries (ckteec
+	PRIVATE pthread
+	PRIVATE teec
+	PRIVATE m
```
Comment (jenswi-linaro) [inline libckteec/CMakeLists.txt]: Why is this needed?

File: libckteec/Makefile [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+$(LIBCKTEEC_OBJ_DIR)/%.o: ${LIBCKTEEC_SRC_DIR}/%.c
+	$(VPREFIX)mkdir -p $(LIBCKTEEC_OBJ_DIR)
+	@echo "  CC      $<"
+	$(VPREFIX)$(CC) $(LIBCKTEEC_CFLAGS) -c $< -o $@
+
+libckteec: $(OUT_DIR)/$(LIBCKTEEC_SO_LIBRARY)
+
+$(OUT_DIR)/$(LIBCKTEEC_SO_LIBRARY): $(LIBCKTEEC_OBJS)
+	@echo "  LINK    $@"
+	$(VPREFIX)$(CC) -shared -Wl,-soname,$(LIBCKTEEC_SO_LIBRARY) $(LIBCKTEEC_LFLAGS) -o $@ $+
```
Comment (jenswi-linaro) [inline libckteec/Makefile]: Where is `LIBCKTEEC_LFLAGS` assigned values?

File: libckteec/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+	PUBLIC include
+	PRIVATE src
+)
+
+target_include_directories(teec
+	PUBLIC include
+)
+
+target_link_libraries (ckteec
+	PRIVATE pthread
```
Comment (jenswi-linaro) [inline libckteec/CMakeLists.txt]: Where libtecc is made we're using `${CMAKE_THREAD_LIBS_INIT}` instead. Any reason why not the same here?
By the way, does the order matter in this list. If so, which should go first?

File: libckteec/src/local_utils.h [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+ * Copyright (c) 2017-2020, Linaro Limited
+ */
+
+#ifndef __LOCAL_UTILS_H
+#define __LOCAL_UTILS_H
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#define LOG_ERROR(...)	printf(__VA_ARGS__)
```
Comment (jenswi-linaro) [inline libckteec/src/local_utils.h]: Any reason why we're not using the `*MSG*()` macros in `public/teec_trace.h` instead?
If you insist on using these macros, you should at least send the output to stderr instead of stdout.

File: libckteec/include/pkcs11_ta.h [hunk 17]
```diff
@@ -0,0 +1,118 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2018-2020, Linaro Limited
+ */
+
+#ifndef PKCS11_TA_H
+#define PKCS11_TA_H
+
+#include <sys/types.h>
+#include <stdint.h>
```
Comment (jenswi-linaro) [inline libckteec/include/pkcs11_ta.h]: I suppose you need `<stdbool.h>` also.

File: libckteec/src/ck_helpers.h [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+#define PKCS11_BAD_PARAM		PKCS11_CKR_ARGUMENTS_BAD
+#define PKCS11_SHORT_BUFFER		PKCS11_CKR_BUFFER_TOO_SMALL
+#define PKCS11_FAILED			PKCS11_CKR_FUNCTION_FAILED
+
+/*
+ * Various conversion between PKCS#11 CK IDs and PKCS11 TA IDs
+ */
+
+/* Conversion from PKCS11 TA to CK only: this is not a one-to-one relationship */
+#define PKCS11_TA_ERROR_CODES \
```
Comment (jenswi-linaro) [inline libckteec/src/ck_helpers.h]: This is a bit horrible, especially if it's one way only. Can we use a more conventional approach in `ta2ck_rv()` instead?
By the way, why is this needed at all? Couldn't the `CKR_*` values be used directly? I assume that the `CKR_*` values are part of the PKCS#11 ABI, so the values are well defined.

File: libckteec/src/ck_debug.c [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+	CK2STR_ENTRY(CKR_PUBLIC_KEY_INVALID),
+	CK2STR_ENTRY(CKR_FUNCTION_REJECTED),
+	CK2STR_ENTRY(CKR_VENDOR_DEFINED),
+};
+
+const char *ckr2str(CK_RV id)
+{
+	static const char vendor[] = "(vendor-defined)";
+	static const char unknown[] = "(unknown)";
+	const int count = sizeof(ckr2str_table) / sizeof(struct ck2str);
```
Comment (jenswi-linaro) [inline libckteec/src/ck_debug.c]: We're bound to need this pattern at other places in this lib. Could we have a common `ARRAY_SIZE()` macro in the utils .h file instead?

File: libckteec/src/invoke_ta.c [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+/*
+ * All requests (invocation of the PKCS11 TA) currently go through a
+ * single GPD TEE session toward the PKCS11 TA.
+ */
+struct ta_primary_context {
+	TEEC_Context context;
+	TEEC_Session session;
+};
+
+static struct ta_primary_context primary_ctx;
```
Comment (jenswi-linaro) [inline libckteec/src/invoke_ta.c]: Do these need to be protected with a mutex?

File: libckteec/src/invoke_ta.c [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+{
+	return (TEEC_Context *)ta_ctx->context;
+}
+
+static TEEC_Session *teec_sess(struct ta_invoke *ta_ctx)
+{
+	return (TEEC_Session *)ta_ctx->session;
+}
+
+TEEC_SharedMemory *ta_alloc_shm(struct ta_invoke *ta_ctx,
```
Comment (jenswi-linaro) [inline libckteec/src/invoke_ta.c]: This symbol will be visible in the exported library, it should have a proper prefix. This comment applies to all other global functions which doesn't use the `ck` prefix.

File: libckteec/src/invoke_ta.h [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+
+#ifndef INVOKE_TA_H
+#define INVOKE_TA_H
+
+#include <pkcs11.h>
+#include <tee_client_api.h>
+
+/*
+ * structure stored in CK session so it calls through the right
+ * GPD TEE session. TEEC type as abstract through void pointers
```
Comment (jenswi-linaro) [inline libckteec/src/invoke_ta.h]: I don't think we need to enforce this internal abstraction. It seems to make things more complicated than needed, especially since we're including `<tee_client_api.h>` any way. So I'd prefer if the proper types where used.

File: libckteec/src/invoke_ta.c [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+{
+	struct ta_invoke *ctx = ta_ctx;
+
+	if (open_primary_context())
+		return NULL;
+
+	if (!ctx)
+		return &primary_invoke;
+
+	if (!ctx->context)
```
Comment (jenswi-linaro) [inline libckteec/src/invoke_ta.c]: This is quite odd. It looks a bit like eighties or nineties application C code to me, where all code was single threaded.
Do we really need this behaviour?

File: libckteec/src/invoke_ta.h [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+ * out_sz is a pointer.
+ *
+ * Function return values versus TA invocation statuses:
+ * - CKR_OK on invocation success (TEEC_SUCCESS)
+ * - CKR_BUFFER_TOO_SMALL on TEEC_ERROR_SHORT_BUFFER. This affect only the
+ *   output data buffer.
+ * - CKR_DEVICE_MEMORY on TEEC_ERROR_OUT_OF_MEMORY.
+ * - and so on...
+ */
+CK_RV ck_invoke_ta(struct ta_invoke *ta_ctx,
```
Comment (jenswi-linaro) [inline libckteec/src/invoke_ta.h]: Here it's called `ta_ctx`, but above it's just `ctx`. Is there some important difference or something else?

File: libckteec/src/pkcs11_api.c [hunk 25]
```diff
@@ -80,24 +83,48 @@ static const CK_FUNCTION_LIST libckteec_function_list = {
 	NOT_SUPPORTED_CK_FUNCTION(C_WaitForSlotEvent),
 };
 
+static bool lib_initiated;
```
Comment (jenswi-linaro) [inline libckteec/src/pkcs11_api.c]: Does this need to be protected by a mutex?

File: libckteec/src/pkcs11_api.c [hunk 26]
```diff
@@ -80,24 +83,48 @@ static const CK_FUNCTION_LIST libckteec_function_list = {
...
+
 CK_RV C_Initialize(CK_VOID_PTR pInitArgs)
 {
-	(void)pInitArgs;
+	CK_C_INITIALIZE_ARGS_PTR args = (CK_C_INITIALIZE_ARGS_PTR)pInitArgs;
 
-	return CKR_FUNCTION_NOT_SUPPORTED;
+	/* Argument currently unused as per the PKCS#11 specification */
```
Comment (jenswi-linaro) [inline libckteec/src/pkcs11_api.c]: If it's unused, why bother with it?

File: libckteec/src/invoke_ta.c [hunk 27]
```diff
@@ -243,6 +243,33 @@ CK_RV ck_invoke_ta(struct ta_invoke *ta_ctx,
...
+	memset(&op, 0, sizeof(op));
+	op.params[2].tmpref.buffer = ta_version;
+	op.params[2].tmpref.size = sizeof(ta_version);
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_NONE, TEEC_NONE,
+					 TEEC_MEMREF_TEMP_OUTPUT, TEEC_NONE);
+
+	res = TEEC_InvokeCommand(sess, PKCS11_CMD_PING, &op, &origin);
+	if (res == TEEC_SUCCESS &&
+	    ta_version[0] == PKCS11_TA_VERSION_MAJOR &&
+	    ta_version[1] == PKCS11_TA_VERSION_MINOR &&
```
Comment (jenswi-linaro) [inline libckteec/src/invoke_ta.c]: `<=` here and for patch.

File: libckteec/src/pkcs11_api.c (hunk 25 above)
Comment (etienne-lms) [reply libckteec/src/pkcs11_api.c]: true.

File: libckteec/src/pkcs11_api.c (hunk 26 above)
Comment (etienne-lms) [reply libckteec/src/pkcs11_api.c]: sorry. I'll fix into
```C
CK_RV C_Initialize(CK_VOID_PTR pInitArgs)
{
	/* Argument currently unused as per the PKCS#11 specification */
	(void)pInitArgs;

	(...)
}
```
keeping inline comment to clarify why the argument is declared but not used.

File: libckteec/include/pkcs11.h (hunk 11 above)
Comment (etienne-lms) [reply libckteec/include/pkcs11.h]: It is built from scratch base on the pkcs11 spec document.

I used a funny indentation here to have all arguments indented the same way in this structure and in the declaration list at line         [line 1002](https://github.com/OP-TEE/optee_client/pull/138/files#diff-a379e8bbffcbc36c341c0bf68b47a88fR1002).

Yet I can move to a more standard way.  I just find it less readable.

File: libckteec/include/pkcs11.h [hunk 28]
```diff
@@ -0,0 +1,1378 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2018-2020, Linaro Limited
+ */
+
+#ifndef __PKCS11_H__
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libckteec/include/pkcs11.h]: yes, will drop.

File: libckteec/include/pkcs11_ta.h (hunk 17 above)
Comment (etienne-lms) [reply libckteec/include/pkcs11_ta.h]: looks like i did not update from https://github.com/OP-TEE/optee_os/pull/2732/ (pkcs11_ta.h)[https://github.com/OP-TEE/optee_os/pull/2732/files#diff-9b89a766b9b62bace834e6ebd6823ea9]. Will do.

File: libckteec/src/ck_debug.c (hunk 19 above)
Comment (etienne-lms) [reply libckteec/src/ck_debug.c]: ok, i'll see that.

File: libckteec/src/ck_helpers.h (hunk 18 above)
Comment (etienne-lms) [reply libckteec/src/ck_helpers.h]: yes, I have this in the pipe. I will upgrade.

By the way, @wamserma [suggested to use `enum`s](https://github.com/OP-TEE/optee_os/pull/2732#discussion_r366766764) for the TA IDs. I answered that I would not like to but in the end I think you're right @wamserma, we should.

Do I move to `enum` all IDs (TA commands, error codes) now, update this P-R and https://github.com/OP-TEE/optee_os/pull/2732, or should a leave that to another P-R once the later is merged?

File: libckteec/src/invoke_ta.c (hunk 20 above)
Comment (etienne-lms) [reply libckteec/src/invoke_ta.c]: should do. I'll fix.

File: libckteec/src/invoke_ta.c (hunk 23 above)
Comment (etienne-lms) [reply libckteec/src/invoke_ta.c]: I don't catch. 
Each call to `C_Initialize()` spawns a TEE session to the TA. Each call to other pkcs11 `C_XxxxXxxx()` API functions leads to invoking the PKCS#11 TA using established TEE session for the related client. Is there something missing?

File: libckteec/src/invoke_ta.c (hunk 21 above)
Comment (etienne-lms) [reply libckteec/src/invoke_ta.c]: ok. i'll use a `ckteec_` for all exported functions.
thanks

File: libckteec/src/invoke_ta.c (hunk 27 above)
Comment (etienne-lms) [reply libckteec/src/invoke_ta.c]: ok.
`==` for major, `<=` for minor and patch.

File: libckteec/src/invoke_ta.h (hunk 22 above)
Comment (etienne-lms) [reply libckteec/src/invoke_ta.h]: ok. I'll try to clean that.

File: libckteec/src/invoke_ta.h (hunk 24 above)
Comment (etienne-lms) [reply libckteec/src/invoke_ta.h]: not. it only aims at referring to the established TA session.

File: libckteec/src/local_utils.h (hunk 16 above)
Comment (etienne-lms) [reply libckteec/src/local_utils.h]: Could use `libteec` macros but maybe one would want `libckteec` traces but not `libteec` traces. No?

File: libckteec/src/invoke_ta.h [hunk 29]
```diff
@@ -0,0 +1,142 @@
...
+ * @shm - memory reference
+ */
+void ta_free_shm(TEEC_SharedMemory *shm);
+
+/**
+ * ck_invoke_ta - Invoke PKCS11 TA for a target request through the TEE
+ *
+ * @ctx - supplied TEE session context
+ * @cmd - PKCS11 TA command ID
+ * @ctrl - command serialized input arguments. Shm, buffer or NULL pointer.
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply libckteec/src/invoke_ta.h]: I guess we can. Will fix.

File: libckteec/src/invoke_ta.c (hunk 27 above)
Comment (jforissier) [reply libckteec/src/invoke_ta.c]: Patch version doesn't matter does it?

Comment (etienne-lms) [reply libckteec/src/invoke_ta.c]: I haven't played yet which changing these numbers, so I cannot tell.
I guess the API should be compliant whatever the patch version number is.

Will drop test on `PKCS11_TA_VERSION_MINOR` if all agree.

Comment (jforissier) [reply libckteec/src/invoke_ta.c]: > Will drop test on PKCS11_TA_VERSION_MINOR if all agree.

Why? For example, if the caller is expecting a feature introduced in 1.2.0 and you get 1.1.0, then it won't work.

Comment (etienne-lms) [reply libckteec/src/invoke_ta.c]: sorry, copy/paste typo, I meant `PKCS11_TA_VERSION_PATCH`.

File: libckteec/src/ck_helpers.h (hunk 18 above)
Comment (jenswi-linaro) [reply libckteec/src/ck_helpers.h]: @wamserma, I'm not sure what you're suggesting. This PR should be OK before it's merged.

**Comment by etienne-lms** [comment]
Hello,
I will submit soon an update of this P-R. Syncing with TA API exported from https://github.com/OP-TEE/optee_os/pull/2733, I must discard few commits about ID conversions originally in this optee_client series. Also I will fully rewrite commit 'libckteec: generic invocation of the PKCS11 TA", not append fixup commit on top of it, since it needed lot of changes.

File: libckteec/src/local_utils.h (hunk 16 above)
Comment (jenswi-linaro) [reply libckteec/src/local_utils.h]: Let's worry about that when it's a problem.

**Comment by jenswi-linaro** [comment]
For commit "libckteec: local utilities" + fixup:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: libckteec/src/invoke_ta.c (hunk 23 above)
Comment (jenswi-linaro) [reply libckteec/src/invoke_ta.c]: I think you got this sorted in the update.

**Comment by jenswi-linaro** [comment]
For commit "libckteec: generic invocation of the PKCS11 TA":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jenswi-linaro** [comment]
For commit "libckteec: sanity on APIs regarding library initialization":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: libckteec/src/invoke_ta.c (hunk 23 above)
Comment (etienne-lms) [reply libckteec/src/invoke_ta.c]: fixed

**Comment by etienne-lms** [comment]
Added late fixup on top of the series for an issue reported by Travis.
Also saw few issues (missing `#include <>` and few commit messages to be updated).
Here is the new series with tags applied where applicable.

File: libckteec/src/invoke_ta.c [hunk 30]
```diff
@@ -5,7 +5,9 @@
  */
 
 /* BINARY_PREFIX is expected by teec_trace.h */
+#ifndef BINARY_PREFIX
```
Comment (etienne-lms) [inline libckteec/src/invoke_ta.c]: I find this very hacky.
Is there's a cleaner way?
Yet, ok to push as is until there's a smarter way.

**Comment by etienne-lms** [comment]
@jenswi-linaro, do you still apply you R-b tag on commit "libckteec: generic invocation of the PKCS11 TA" with the 2 fixup commit squashed?

**Comment by jenswi-linaro** [comment]
> @jenswi-linaro, do you still apply you R-b tag on commit "libckteec: generic invocation of the PKCS11 TA" with the 2 fixup commit squashed?

Yes, that's OK.

File: libckteec/include/pkcs11.h (hunk 11 above)
Comment (jenswi-linaro) [reply libckteec/include/pkcs11.h]: I'm sure that checkpatch would have a thing or two to say about this.

**Comment by jenswi-linaro** [comment]
Please rebase on master to include checkpatch in the review.

**Comment by etienne-lms** [comment]
Update for checkpatch compliance. I squashed all fixup related to checkpatch complains in "libckteec: Introduce PKCS#11 API (2.40-e01) header …".

I see there are still failures. Most I'll fix. But there ar esome I don't understand what wring with:
* `ARRAY_SIZE`. What wrong with the label?
```
total: 0 errors, 0 warnings, 1 checks, 96 lines checked
47f5399 libckteec: local utilities
WARNING: Prefer ARRAY_SIZE(array)
#29: FILE: libckteec/src/local_utils.h:9:
+#define ARRAY_SIZE(array)	(sizeof(array) / sizeof(array[0]))
```
* `BIT()` macro. There is no standard header file for `BIT()` hence I prefer to not rely on `BIT` in the user client API header file.
```
total: 0 errors, 0 warnings, 3 checks, 1020 lines checked
a86e8c1 libckteec: Introduce PKCS#11 API (2.40-e01) header
CHECK: Prefer using the BIT macro
#158: FILE: libckteec/include/pkcs11.h:93:
+#define CKF_ARRAY_ATTRIBUTE		(1U << 30)
```

**Comment by jenswi-linaro** [comment]
I'm fine with ignoring the `BIT()` and `ARRAY_SIZE()` warnings.

**Comment by jenswi-linaro** [comment]
Let me know when you're done with the current updates and I'll go through it once more.

**Comment by etienne-lms** [comment]
Another fixup commit for "libckteec: implement C_GetFunctionList()" (R-b tagged). https://github.com/OP-TEE/optee_client/pull/138/commits/b16330bb94b91ba7283a681d847f96e04663798a simplifies the code the comply with checkpatch.
I will squash the 2 fixup commits (https://github.com/OP-TEE/optee_client/pull/138/commits/b16330bb94b91ba7283a681d847f96e04663798a and https://github.com/OP-TEE/optee_client/pull/138/commits/a9bc912289cb25fc502e0ee430942a9649c1aa2d) if you agree.

**Comment by etienne-lms** [comment]
I am done with the updates.

File: libckteec/src/pkcs11_api.c [hunk 31]
```diff
@@ -6,82 +6,79 @@
...
-	NOT_SUPPORTED_CK_FUNCTION(C_SeedRandom),
-	NOT_SUPPORTED_CK_FUNCTION(C_GenerateRandom),
-	NOT_SUPPORTED_CK_FUNCTION(C_GetFunctionStatus),
-	NOT_SUPPORTED_CK_FUNCTION(C_CancelFunction),
-	NOT_SUPPORTED_CK_FUNCTION(C_WaitForSlotEvent),
+	.C_Initialize = C_Initialize,
+	.C_Finalize = C_Finalize,
+	.C_GetInfo = NULL,
+	.C_GetFunctionList = C_GetFunctionList,
+	.C_GetSlotList = NULL,
```
Comment (jenswi-linaro) [inline libckteec/src/pkcs11_api.c]: I'm not sure we need all these NULL assignments, it would be done any way if skipped.
As documentation on which function pointer exists it's better to rely on the definition of `CK_FUNCTION_LIST`.

Comment (etienne-lms) [reply libckteec/src/pkcs11_api.c]: true, I'll do. thanks.

**Comment by jenswi-linaro** [comment]
Looks good to me.
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
squashed. tags applied.

**Comment by jbech-linaro** [comment]
FYI, I've noticed that this PR breaks the `make optee-client-common` in build.git. Haven't checked if it is just a missing flag or if it's actually broken. Looks like CMake files have been updated, but not the GNU make files.
```bash
$ cd build && make -j`nproc` optee-client-common                                         
make -C /home/jbech/devel/optee_projects/qemu/build/../optee_client CROSS_COMPILE="/usr/bin/ccache /home/jbech/devel/optee_projects/qemu/build/../toolchains/aarch32/bin/arm-linux-gnueabihf-
" CFG_TEE_BENCHMARK=n CFG_TA_TEST_PATH=y          
make[1]: Entering directory '/home/jbech/devel/optee_projects/qemu/optee_client'
Building libteec.so                                                     
Building libckteec.so                 
  CC      src/ck_helpers.c                                     
  CC      src/pkcs11_api.c                                 
  CC      src/invoke_ta.c
  CC      src/tee_client_api.c                                          
  CC      src/teec_trace.c                                                
  LINK    /home/jbech/devel/optee_projects/qemu/optee_client/libckteec/../out/libckteec/libckteec.so.0.1.0
  AR      /home/jbech/devel/optee_projects/qemu/optee_client/libckteec/../out/libckteec/libckteec.a
/home/jbech/devel/optee_projects/reference/toolchains/aarch32/bin/../lib/gcc/arm-linux-gnueabihf/8.3.0/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lteec
collect2: error: ld returned 1 exit status                            
Makefile:47: recipe for target '/home/jbech/devel/optee_projects/qemu/optee_client/libckteec/../out/libckteec/libckteec.so.0.1.0' failed
make[2]: *** [/home/jbech/devel/optee_projects/qemu/optee_client/libckteec/../out/libckteec/libckteec.so.0.1.0] Error 1
Makefile:40: recipe for target 'build-libckteec' failed                  
make[1]: *** [build-libckteec] Error 2     
make[1]: *** Waiting for unfinished jobs....                     
  AR      /home/jbech/devel/optee_projects/qemu/optee_client/libteec/../out/libteec/libteec.a
  LINK    /home/jbech/devel/optee_projects/qemu/optee_client/libteec/../out/libteec/libteec.so.1.0.0
                                                       
make[1]: Leaving directory '/home/jbech/devel/optee_projects/qemu/optee_client'
common.mk:413: recipe for target 'optee-client-common' failed
make: *** [optee-client-common] Error 2

```

**Comment by jforissier** [comment]
@jbech-linaro I've noticed that too, does not happen every time. Now I have double checked, it's just a missing dependency. I'll create a PR.

**Comment by jforissier** [comment]
https://github.com/OP-TEE/optee_client/pull/183
